src/BloomExe/ImageProcessing/RuntimeImageProcessor.cs

```diff
@@ -0,0 +1,201 @@
+					{
+						_originalPathToProcessedVersionPath.Remove(originalPath);
+					}
+				}
+
+				using (var originalImage = PalasoImage.FromFile(originalPath))
+				{
+					//if it's a jpeg, we don't resize, we don't mess with transparency, nothing. These things
+					//are scary in .net. Just send the original back and wash our hands of it.
+					if (ImageUtils.AppearsToBeJpeg(originalImage))
+					{
+						return originalImage.OriginalFilePath;
+					}
+
+					double shrinkFactor = 1.0;
+
+#if ShrinkLargeImages // at the moment, we're suspecting that it may be better to let the browser do the shrinking
+
+					//if its a small image, like a creative commons logo, we don't try and resize it
+					if (originalImage.Image.Width > TargetDimension || originalImage.Image.Height > TargetDimension)
+					{
+						var maxDimension = Math.Max(originalImage.Image.Width, originalImage.Image.Height);
+						//enhance: if we had a way of knowing what the target dimension actually was, we'd use that, of course
+						shrinkFactor = (TargetDimension/(double) maxDimension);
+					}
+#endif
+					var destWidth = (int) (shrinkFactor*originalImage.Image.Width);
+					var destHeight = (int) (shrinkFactor*originalImage.Image.Height);
+
+					using (var processedBitmap = new Bitmap(destWidth, destHeight))
+					{
+						using (Graphics g = Graphics.FromImage((Image) processedBitmap))
+						{
+							//in version 1.0, we used .NearestNeighbor. But if there is a border line down the right size (as is common for thumbnails that,
+							//are, for example, re-inserted into Teacher's Guides), then the line gets cut off. So I switched it to HighQualityBicubic
+							g.InterpolationMode = InterpolationMode.HighQualityBicubic; //.NearestNeighbor;//or smooth it: HighQualityBicubic
+							var destRect = new Rectangle(0, 0, destWidth, destHeight);
+							lock (_transparentImageAttributes)
+							{
+								g.DrawImage(originalImage.Image, destRect, 0, 0, originalImage.Image.Width, originalImage.Image.Height,
+									GraphicsUnit.Pixel, _transparentImageAttributes);
+							}
+						}
+
+						var pathToProcessedImage = Path.Combine(_cacheFolder, Path.GetRandomFileName() + Path.GetExtension(originalPath));
+
+						//Hatton July 2012:
+						//Once or twice I saw a GDI+ error on the Save below, when the app 1st launched.
+						//I verified that if there is an IO error, that's what you get (a GDI+ error).
+						//I looked once, and the %temp%/Bloom directory wasn't there, so that's what I think caused the error.
+						//It's not clear why the temp/bloom directory isn't there... possibly it was there a moment ago
+						//but then some startup thread cleared and deleted it? (we are now running on a thread responding to the http request)
+
+						Exception error = null;
+						for (int i = 0; i < 5; i++) //try up to five times, a second apart
+						{
+							try
+							{
+								error = null;
+
+								if (!Directory.Exists(Path.GetDirectoryName(pathToProcessedImage)))
```

Comment:
It's safe to just call CreateDirctory unconditionally (does nothing if it exists).